### PR TITLE
Feature/proc hackathon

### PR DESCRIPTION
**Pull request by corentincarton**
### Summary
This PR introduces a broader, more consistent module structure across earthkit-meteo, centered on a flexible dispatch approach for high-level APIs and clearer backend separation (array, xarray, and fieldlist in the future).

It aligns API design, documentation style, and test organization across major modules (wind, extreme, stats, solar, thermo, regimes, score).

### Main Changes
**1. Dispatch infrastructure and API consistency**

- Added/extended a reusable dispatch system in earthkit.meteo.utils.decorators.
- Updated high-level functions in multiple modules to dispatch to backend implementations in a consistent way.
- Improved dispatch error handling and flexibility (matching args/kwargs and optional array dispatch).

**2. Backend structure expansion (array / xarray / fieldlist)**

- Added new xarray backends in several modules (notably wind, extreme, solar, stats, thermo).
- Added/updated fieldlist backend support where needed (notably wind low-level/backend).
- Harmonized high-level module entry points and backend imports.

**3. Wind module restructuring**
Added dedicated backend modules, typically
- wind/array
- wind/xarray
- wind/fieldlist
High-level wind API now dispatches to array/xarray (FieldList remains backend/experimental-oriented).

**4. Extreme/stats/regimes/solar/thermo/score alignment**

- Added high-level + backend split for extreme indices and related stats utilities.
- Added xarray-compatible implementations for several modules.
- Unified dimension-handling patterns and docstrings across similar APIs.

**5. Tests and structure alignment**

- Reorganized tests to better mirror module/backend structure:
- e.g. separate *_array and *_xarray suites
- Added high-level compatibility tests for dispatch behavior.
- Added/updated tests for edge cases and backend parity.

**6. Docs and project config**

- Updated docs config/rules and examples.
- Updated dependencies and test requirements where needed for xarray/scipy-backed features.

### Notable Behavior / Compatibility Notes

- High-level dispatch behavior is now explicit and more uniform across modules.
- In wind, high-level FieldList usage is intentionally de-emphasized compared to array/xarray dispatch as the FieldList interface will be implemented later across the modules.
- Tests have been split/renamed in several areas to reflect backend-specific structure.

### Validation

- Added/updated unit tests across affected modules (wind, extreme, stats, solar, thermo, regimes, score).
- Backend-specific suites (array/xarray/fieldlist) now provide clearer coverage by layer.

### Contributor Declaration

By opening this pull request, I affirm the following:

* All authors agree to the [Contributor License Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).
* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.

### CODE REVIEW

**Comment by corentincarton** [comment]
@claude[agent]  please perform a full PR review of the new `earthkit-meteo` package structure, with a focus on architecture consistency, dispatch-based APIs, documentation, and tests.
 
Review goals:
1. Dispatch docstring consistency
- Find all functions using the dispatch approach (e.g. via `dispatch(...)` in high-level modules).
- Ensure each such function docstring ends with exactly:
  "The function returns an object of the same type as the input arguments."
- List missing/inconsistent cases and propose exact edits.
 
2. Test layout consistency with module layout
- Check that test folders/files mirror module structure.
- Target structure example:
  - `tests/wind/test_wind.py`
  - `tests/wind/array/test_wind.py`
  - `tests/wind/xarray/test_wind.py`
- Identify modules where tests do not mirror source layout and propose a concrete migration plan (file moves/renames and rationale).
 
3. Cross-module alignment audit
- Verify modules follow consistent:
  - package/file structure
  - API layering (high-level vs backend-specific modules)
  - dispatch usage patterns
  - type hints and overload patterns
  - docstring format/style
  - test strategy and naming conventions
- Compare at least wind + a few other representative modules (e.g. extreme, stats, vertical) and call out mismatches.
 
4. Improvement suggestions
- Provide prioritized recommendations for alignment (high/medium/low priority).
- Include actionable changes with file paths and brief examples.
- Flag any risks/regressions introduced by structural differences.
 
Output format:
- **Findings** (ordered by severity)
- **Gaps vs expected structure**
- **Proposed changes** (specific and actionable)
- **Optional follow-ups** (nice-to-have refactors)
 
Be strict and concrete: no generic comments. Include exact paths and suggested wording where relevant.

**Comment by Oisin-M** [comment]
Current test failures
- [ ] `FAILED tests/stats/test_stats.py::test_quantiles_nans[arr0-array_api_compat.torch-device1] - ValueError: only one element tensors can be converted to Python scalars`
- [ ] `FAILED tests/stats/test_stats.py::test_quantiles_core[data0-4-kwargs0-v_ref0-sort-array_api_compat.torch-device1] - AssertionError: i=0, d=tensor([[0.0000, 0.0000, 0.0000, 0.0000],`

To be checked
- [ ] device handling - can we run with torch on cuda correctly?

File: src/earthkit/meteo/wind/xarray/wind.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        ``speed`` are histogrammed along the first dimension and values in ``direction``
+        are histogrammed along the second dimension.
+    xarray.DataArray
+        The direction bins (i.e. the sectors) (degrees)
+
+
+    Notes
+    -----
+    The ``sectors`` parameter defines the number of direction bins the 360 degreesThe sectors do not start at 0 degrees (North) but are shifted by half a sector size.
+    E.g. if ``sectors`` is 4 the sectors are defined as:
```
[the remark this answers is not in the document]
Comment (corentincarton) [reply src/earthkit/meteo/wind/xarray/wind.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/ecmwf/earthkit-meteo/pull/119#discussion_r2924464676)

File: src/earthkit/meteo/wind/xarray/wind.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    dir_bins = np.asarray(dir_bins, dtype=res.dtype)
+
+    res_da = xr.DataArray(
+        res,
+        dims={"speed_bin": res.shape[0], "direction_bin": res.shape[1]},
+        coords={
+            "speed_bin": speed_bins[:-1],
+            "direction_bin": dir_bins[:-1],
+        },
+    )
```
[the remark this answers is not in the document]
Comment (corentincarton) [reply src/earthkit/meteo/wind/xarray/wind.py:302]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/ecmwf/earthkit-meteo/pull/119#discussion_r2924464771)

**Comment by sandorkertesz** [comment]
I noticed that `apply_ufunc` appears in the docs for solar: https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/solar/xarray/index.html

**Comment by sandorkertesz** [comment]
Some docs issues:
- formatting: https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/solar/incoming_solar_radiation.html#earthkit.meteo.solar.incoming_solar_radiation
- formatting: https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/wind/windrose.html#earthkit.meteo.wind.windrose
- thermo: does not seem to use "overload" like "wind" or "solar". So top level function docs leads to the "array" one

**Comment by corentincarton** [comment]
> I noticed that `apply_ufunc` appears in the docs for solar: https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/solar/xarray/index.html

Good catch @sandorkertesz. The current import model is too simple and imports everything in most module. It's done correctly in extreme. I'll fix that using that approach.

**Comment by sandorkertesz** [comment]
Some issues in the docs in `thermo`:
- Array level is completely missing:  https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/thermo/kelvin_to_celsius.html#earthkit.meteo.thermo.kelvin_to_celsius
- Array level is missing from "Implementations": https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/thermo/lcl.html#earthkit.meteo.thermo.lcl
- Wrongly formatted formulas (many) e.g.:
      - https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/thermo/lcl_temperature.html
      - https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/thermo/saturation_specific_humidity_slope.html

**Comment by corentincarton** [comment]
> * https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/thermo/kelvin_to_celsius.html#earthkit.meteo.thermo.kelvin_to_celsius

@sandorkertesz I just pushed a version with the array API at the high-level for thermo... not sure how we forgot that one! I'll have a look at the formulas later.

**Comment by sandorkertesz** [comment]
Thanks! All looks fine except the vertical documentation: https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/vertical/index.html Here the Xarray documentation is missing (the array-level and Xarray level methods are different for vertical,  `interpolate_monotonic` being the only shared method).

**Comment by Oisin-M** [comment]
> Thanks! All looks fine except the vertical documentation: https://earthkit-meteo--119.org.readthedocs.build/en/119/autoapi/earthkit/meteo/vertical/index.html Here the Xarray documentation is missing (the array-level and Xarray level methods are different for vertical, `interpolate_monotonic` being the only shared method).

I reorganised things so now we can dispatch properly, and both array and xarray are showing in documentation. One problematic method however is `hybrid_level_parameters` because this does not take as input an array and therefore does not fit into the current dispatching assumptions.

Second thing to note is that docstrings are missing from the toplevel. Previously everything was *args, **kwargs so I at least updated to have the correct signatures (required for dispatch to work), but haven't done anything else.

And lastly, `vertical_axis` is used in the array implementations, but we probably want to make this `vertical_dim` for consistency.

**Comment by corentincarton** [comment]
@sandorkertesz, shall we merge this soon? What is missing?

**Comment by sandorkertesz** [comment]
@corentincarton, it can be merged. But the `vertical_axis` has to be changed before it is released.

**Review by sandorkertesz — APPROVED** [review]
